enable relaxed constexpr in core

gnzlbg opened the pull request:
- adds constexpr to:
  - begin, end, rbegin, rend, cbegin, cend, crbegin, crend
    - note: i've hijacked ADL for C-arrays and initializer lists in ranges::rbegin/rend/crbegin/crend 
      because std::rbegin/rend/crbegin/crend are not constexpr (while std::begin/std::end are...). 
        - a constexpr reverse_iterator is added to utility to make this work.
  - advance, next, prev, distance, iter_size, swap, indirect_swap
    - note 1: i've hijacked std::advance for pointers since it is not constexpr
    - note 2: for the test iterators I have to define advance overloads that forward
      to range-v3 ones. I think it would be better to allow customizing advance's implementation
      via a different name (e.g. advance_customization_point), and not import std::advance at all. 
  - front, back, empty, size, at

- adds constexpr tests to test/constexpr_core.cpp
   - test iterators are constexpr
   - test::array is fully tested, C-arrays and initializer_lists are partially tested (since swap for them
     is not constexpr and iter_swap cannot work). view::iota tests can be added in the future.

- adds reverse_iterator to utilities
- adds testing utility array
- fix: remove c++1y test for view::join, it is already tested in C++14 mode.

ericniebler commented on include/range/v3/utility/addressof.hpp:
```diff
@@ -0,0 +1,78 @@
+            {
+                return std::addressof(r);
+            }
+        };
```
Ogh. This is exactly the sort of ugly hack expression SFINAE *should* solve. Is there no better way?

ericniebler replied:
So, you can just use `reinterpret_cast` here because that's not `constexpr`?

gnzlbg replied:
> So, you can just use reinterpret_cast here because that's not constexpr?

Yes, but then I would be reimplementing `std::addressof` which depending on the std library might be doing a bit more than that (e.g. libc++ interfaces with Objective-C++ pointers...).

gnzlbg replied:
It might be worth it to also tests here if the user-provided addressof operator is constexpr or not.

ericniebler commented on include/range/v3/utility/array.hpp:
```diff
@@ -0,0 +1,304 @@
+
+}  // namespace std
+
+#endif // RANGES_V3_UTILITY_ARRAY_HPP
```
Oh, no. No no. If `constexpr` support means wholesale duplication of std functionality, I don't want it. This has to go.

gnzlbg replied:
I guess you mean the duplication of `std::array` in general (correct me if I'm wrong). 

The duplication is not necessary, but it simplifies testing so much that it is worth having. As stated in the pull request, I'm happy to move this to the test directory. The issue it solves is that besides C-arrays, and initializer lists, I need to test against a constexpr friendly user defined sequence. This array class is way simpler than `view::ints` to get right, and adds almost no dependencies. At the beginning I was using `view::ints` to test everything, but that pulls more of range-v3 into each test and requires a lot of `view`-related functionality to be constexpr. I didn't want to submit all those changes just yet. I've done it once, and I prefer to tackle this step by step, in the following order: core, algorithms, views, and finally actions.

So I guess I agree, it sucks that `std::array` is broken, otherwise I would use it. But this is better than the other alternatives I've tried, so I would like to keep it.

ericniebler replied:
I would suggest a minimal constexpr-friendly `array` container in the test directory. Containers don't belong in range-v3 proper.

gnzlbg replied:
Done.

ericniebler commented on include/range/v3/utility/reverse_iterator.hpp:
```diff
@@ -0,0 +1,203 @@
+
+    }  // namespace v3
+}  // namespace ranges
+#endif  // RANGES_V3_UTILITY_REVERSE_ITERATOR_HPP
```
Some things:

1) I *really* don't like duplicating std functionality. I've done in in isolated places (e.g. some std iterators weren't modeling the Iterator concepts of the Palo Alto report), but I did it grudgingly.
2) The iterator adaptors live in utility/iterator.hpp
3) Most of this verbiage can be replaced with an appropriately-defined cursor together with the `basic_iterator` utility.
4) Don't use `iterator_traits`. New-style iterators don't define nested `pointer` or `reference` type aliases or `operator->`.

ericniebler replied:
Ah. If you remove `operator->` then you can nuke the weird `ranges::addressof`. Cool.

gnzlbg replied:
> 1) I really don't like duplicating std functionality. I've done in in isolated places (e.g. some std iterators weren't modeling the Iterator concepts of the Palo Alto report), but I did it grudgingly.

I don't like it either.

> 2) The iterator adaptors live in utility/iterator.hpp
> 3) Most of this verbiage can be replaced with an appropriately-defined cursor together with the basic_iterator utility.

This is libc++'s `reverse_iterator` implementation. I'll write a reverse iterator using `basic_iterator` and a cursor and make sure it passes all libc++ tests. And put that into `utility/iterator.hpp`.

> 4) Don't use iterator_traits. New-style iterators don't define nested pointer or reference type aliases or operator->.
> Ah. If you remove operator-> then you can nuke the weird ranges::addressof. Cool.

Cool, didn't knew about this. I'll remove those then (`addressof` will come back for variant tho).

ericniebler replied:
> I'll write a reverse iterator using basic_iterator and a cursor and make sure it passes all libc++ tests.

Obviously, if the libc++ tests check for the presence of `operator->` or whatever, those tests will fail and that's ok. Some simple sanity checks would do IMO.

gnzlbg replied:
@ericniebler I've mostly done this in the last commit, and am pretty amazed at how easy it was to do it with cursor and basic iterator (even without docs). 

There are only two minor quirks missing. One is a warning during "concept checking" that occurs in the test constexpr_core.cpp. The other is a situation that used to compile with `std::reverse_iterator` but doesn't compile with the `basic_iterator` implementation. It is related to  `reverse_iterator<non_const_it> - reverse_iterator<const_it>`, where  `reverse_iterator<const_it>` has to be converted to `reverse_iterator<non_const_it>` when using `basic_iterator`, which fails.

ericniebler commented on include/range/v3/utility/swap.hpp:
```diff
@@ -42,10 +53,26 @@ namespace ranges
+                T t(std::move(x));
+                x = std::move(y);
+                y = std::move(t);
+            }
```
`std::move` isn't `constexpr` in C++11.

ericniebler replied:
This just doesn't work. It's perfectly acceptable for users to explicitly specialize `swap` on user-defined types in namespace `std`. Routing around `std::swap` just because it happens not to be `constexpr` is fundamentally broken.

ericniebler replied:
...And it seems to me that if you try to call `swap` on a type that happens to have `std` as an associated namespace, it's going to find the one in `std` anyway. This overload creates an ambiguity error, does it not?

gnzlbg replied:
Thanks for taking a look at this. I didn't knew users could specialize `std::swap` in the `std` namespace. I thought the preferred way was to add a swap function to the namespace of the user-defined type and let ADL do the rest. If explicit specializations of `std::swap` are to be considered, then I need to reconsider this approach (more below).

First, what I want and tried to do is to let ADL pick the swap function in the namespace of the user-defined type. For literal types, or if no swap function is found, it should pick this swap function and not the one in namespace std. As far as I can tell, this overload does not create any ambiguity. It would if I were to import `std::swap`. 

If I have to allow for explicit specializations of std swap, then I need to import std::swap. I don't see a way to import those without importing `std::swap`. The whole thing is not doomed tho, since I can probably add a constexpr swap to the ranges namespace that will be picked for types living on it, and users can add constexpr swap functions for their own types if desired. However, I don't see a way to add a constexpr swap for C-arrays if I have to import the std::swap here (I've just realized that I forgot to add the swap overload for C-arrays that calls `swap_ranges` but I can fix this.). What do you think?
 
> std::move isn't constexpr in C++11.

The function is only constexpr in C++14 or greater.

gnzlbg replied:
So I just learned that yes, it is common to explicitly specialize std::swap since you cannot use ADL for hooking up a swap for e.g. `std::reference_wrapper<T>`. I'll rethink this.

ericniebler replied:
I think you're mistaken about the ambiguity error. I just tried the following with your branch:

~~~~cpp
#include <utility>
#include <complex>
#include <range/v3/core.hpp>

template<typename T>
struct S {
    T t;
};

int main()
{
    S<std::complex<float>> s,t;
    ranges::swap(s,t);
}
~~~~

With clang, this is the error I get:

~~~
/cygdrive/c/Users/eric/Code/range-v3/libs/range/v3/scratch/range.cpp:13:5: error: no matching function for call to object of type 'const ranges::v3::adl_swap_detail::swap_fn'
    ranges::swap(s,t);
    ^~~~~~~~~~~~
/cygdrive/c/Users/eric/Code/range-v3/include/range/v3/utility/swap.hpp:94:17: note: candidate template ignored: substitution failure [with T = S<std::complex<float> > &, U = S<std::complex<float> > &]: no type named 'type' in
      'meta::v1::detail::_if_<std::integral_constant<bool, false> >'
                operator()(T && t, U && u) const noexcept(is_nothrow_swappable<T, U>::value)
                ^
1 error generated.
~~~

With my branch, it compiles fine. Reread my comment above about types with `std` as an associated namespace and ADL. This is the major problem with your approach. I'm quite certain that any effort to redefine `swap` will fail this way.

gnzlbg replied:
You are right. I've reverted the swap changes, and added this test to `test/utility/swap.cpp`.

After sleeping over it, I think the best thing is to just import std swap. Since ADL is used users can implement constexpr versions of swap for their own types if so desired. 

The only case that might be worth considering is that of C-arrays, but that can be done in the `swap_fn` object before performing ADL.

gnzlbg replied:
I worked out a new implementation that defaults to ADL and falls back to std::swap. 

Users can decide if they want ADL + std::swap, ADL + ranges::swap, or just ranges::swap without ADL (this is useful for std library types).

ericniebler replied:
Please just put swap back the way it was.

gnzlbg replied:
Done. 

I've had to remove the constexpr `iter_swap` and `swap` tests since `std::swap` will be picked up for, among others, fundamental types. The associated namespaces of fundamental types are empty, and there is not much that can be done about this. As a consequence, e.g. `iter_swap(int*, int*)` won't be constexpr. I don't know how large of a hit this will be, but I guess it will be a pretty large hit (see the commented out code in the `constexpr_core.cpp` tests). One can work around this for a user defined type (e.g. see `test/array.hpp`), but that workaround is just wrong (I just left it for exposition only).

ericniebler replied:
I understand. We can revisit this in the future.

ericniebler commented:
> maybe importing `rbegin`/`rend` should be removed, since i've hijacked them anyways.

A decision needs to be made about whether `rbegin`/`rend` are customization points. If they are, we can't remove the `using` declarations you refer to without causing potential ambiguity errors.

gnzlbg commented:
> A decision needs to be made about whether rbegin/rend are customization points.

I'm inclined to think that they are customization points. Until now I always asked myself the question "Does it make sense to customize this function for C-arrays?". Maybe this is the wrong question, but for begin, end, size, rbegin, and rend the answer is yes. If it makes sense for C-arrays, it might make sense for some other user-defined type that one cannot easily modify. 

>If they are, we can't remove the using declarations you refer to without causing potential ambiguity errors.

Then they stay. I've only "hijacked" them for C-arrays and `std::initializer_lists`.

ericniebler commented on include/range/v3/utility/swap.hpp:
```diff
@@ -23,6 +34,14 @@
+    b = tmp;
+}
+
+
```
Remove this.

gnzlbg replied:
Oops, fixed.

ericniebler commented on include/range/v3/utility/iterator.hpp:
```diff
@@ -483,25 +526,119 @@ namespace ranges
+        template<typename I>
+        using reverse_iterator = basic_iterator<detail::reverse_cursor<I>,
+                                                detail::reverse_cursor<I>>;
+
```
These belong in range_fwd.hpp

gnzlbg replied:
Fixed.

ericniebler commented on include/range/v3/utility/iterator.hpp:
```diff
@@ -483,25 +526,119 @@ namespace ranges
+
+        template<typename I>
+        struct reverse_cursor
+        {
```
Formatting

gnzlbg replied:
Fixed.

gnzlbg commented:
> Also, the fact that basic_iterator still provides operator-> (and hence needs std::addressof) is merely an oversight. It ought to be deleted.
> 4) Don't use iterator_traits. New-style iterators don't define nested pointer or reference type aliases or operator->.

I've removed `operator->`, `pointer`,  and also `pointer` from the specialization of `std::iterator_traits<basic_iterator<C, S>>`, but removing reference gives too many errors. Having a crippled `std::iterator_traits` is not something anybody will expect.

I fixed the conversion tests that were not passing for `reverse_iterator`. The fix is not finished tho since I still need to constrain the `basic_iterator::operator-(other_basic_iterator)` more. The issue was:

```c++
char* a;
const char * b;
// this worked:
b - a;
// since it expands to b.operator-(a)
// and a (char*) can be converted to b (const char*)
// this failed:
a - b; 
// since b (const char*) cannot be converted to a (char*)
```

I cannot constraint operator- with `Convertible` since the types are not convertible to each other, but they can be substracted from each other. I also don't know which difference_type to choose, probably the widest of both is the right solution.



t break

ericniebler commented:
>> Also, the fact that basic_iterator still provides operator-> (and hence needs std::addressof) is merely an oversight. It ought to be deleted.
>> 4) Don't use iterator_traits. New-style iterators don't define nested pointer or reference type aliases or operator->.
>
> I've removed operator->, pointer, and also pointer from the specialization of std::iterator_traits<basic_iterator<C, S>>, but removing reference gives too many errors. Having a crippled std::iterator_traits is not something anybody will expect.

Sorry, I meant the *iterators* don't need nested `pointer` or `reference` type aliases. `iterator_traits` still needs them, though.

> I fixed the conversion tests that were not passing for reverse_iterator. The fix is not finished tho since I still need to constrain the basic_iterator::operator-(other_basic_iterator) more. The issue was:
> 
> ~~~~
> char* a;
> const char * b;
> // this worked when using 
> // basic_iterator<random_access_iterator<char*>> and ...<<const char*>>:
> b - a;
> // since it expands to b.operator-(a)
> // and a (char*) can be converted to b (const char*)
> // this failed:
> a - b; 
> // since b (const char*) cannot be converted to a (char*)
> ~~~~
>
> I cannot constrain operator- with Convertible since the types are not convertible to each other, but they can be substracted from each other. I also don't know which difference_type to choose, probably the widest of both is the right solution.

I'll see what can be done about that.

gnzlbg commented on include/range/v3/range_access.hpp:
```diff
@@ -162,19 +162,19 @@ namespace ranges
-            template<typename Cur>
-            static auto distance_to(Cur const &pos0, Cur const &pos1)
+            template<typename Cur0, typename Cur1>
+            static RANGES_CXX14_CONSTEXPR auto distance_to(Cur0 const &pos0, Cur1 const &pos1)
```
I had to change this to allow for computing the distance between iterators of different types that are not convertible to each other but can still be "substracted" (`const char*`, `char`).

ericniebler replied:
I fixed this problem in `basic_iterator`. Apologies for the merge conflicts. :-(

gnzlbg replied:
Thanks! No big deal and everything works fine now :)

ericniebler commented on include/range/v3/utility/basic_iterator.hpp:
```diff
@@ -411,15 +436,18 @@ namespace ranges
-            Cur &pos() noexcept
+
+          public:
+            RANGES_CXX14_CONSTEXPR Cur &pos() noexcept
```
`pos` should not be public.

ericniebler commented on include/range/v3/utility/basic_iterator.hpp:
```diff
@@ -709,7 +754,6 @@ namespace std
                 ::ranges::detail::as_std_iterator_category<
                     typename iterator::iterator_category,
                     reference>>;
-        using pointer = typename iterator::pointer;
```
This line needs to be restored.

ericniebler commented on include/range/v3/utility/iterator.hpp:
```diff
@@ -480,25 +523,111 @@ namespace ranges
+                template<typename U, CONCEPT_REQUIRES_(Convertible<I, U>{})>
+                RANGES_CXX14_CONSTEXPR operator reverse_cursor<U>() {
+                    return U{it_};
+                }
```
You're liable to get conversion ambiguities if you have both a converting constructor and a conversion operator. Remove the conversion operator and keep the constructor.

ericniebler commented on include/range/v3/utility/iterator.hpp:
```diff
@@ -480,25 +523,111 @@ namespace ranges
+                distance_to(reverse_cursor<U> const &that) const
+                {
+                    return it_ - that.base();
+                }
```
This doesn't need to be a template. The machinery in `basic_iterator` makes sure the comparison is symmetric by forcing a conversion.

ericniebler commented on include/range/v3/utility/swap.hpp:
```diff
@@ -10,6 +10,17 @@
+// This file is dual licensed under the MIT and the University of Illinois Open
+// Source Licenses. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
```
Remove this comment.

ericniebler commented:
After you've finished with those last few edits, can you rebase your pull request onto master and squash all your commits into one? [Instructions](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request)

ericniebler commented:
Looks very good, thanks.
